### PR TITLE
gsoc26: Mapping Conversion Layer (Layer 3), #60

### LINKED ISSUE #60: gsoc26: Mapping Conversion Layer (Layer 3)
# Feature Request

**Describe the feature you'd like**
Add Layer 3 (cross-class mapping conversion) to the Databus Python Client download pipeline. Users should be able to convert downloaded RDF files to CSV and vice versa, and convert between RDF triple and quad formats, using the existing `--convert-format` flag extended with `--graph-name` and `--base-uri` flags.

**Why is this feature important?**
Many applications consume tabular data (CSV/TSV) but datasets on the Databus are published in RDF formats. Cross-class conversion eliminates the manual mapping step and makes Databus datasets directly consumable by tabular applications. It also enables named graph workflows by supporting triples-to-quads promotion and quads-to-triples splitting.

**Describe alternatives you've considered**
The Java client uses TARQL for CSV to RDF mapping. For the Python client, a direct rdflib-based approach is used instead to avoid adding a new dependency. The companion metadata JSON file approach is used to address data loss of RDF datatypes and language tags during RDF to CSV conversion, enabling lossless round trips when both files are used together.

**Additional context**
Supported mapping directions:

| From | To | Notes |
|---|---|---|
| RDF Triples | RDF Quads | Requires `--graph-name`. Lossless. |
| RDF Quads | RDF Triples | Splits into one file per named graph. Lossless. |
| RDF Triples | CSV | Quasi-equal. Companion .meta.json generated. |
| CSV | RDF Triples | Lossless if companion file present. Requires `--base-uri`. |

Out of scope for this issue:
- RDF Quads to CSV (stretch goal, significant added complexity)
- TARQL integration

### PR DESCRIPTION
**Pull request by DhanashreePetare**
# Pull Request

## Description

Implements Layer 3 (cross-class mapping conversion) for the Databus Python Client download pipeline, building on the Layer 2 format handlers (TripleHandler, QuadHandler, TSDHandler) merged in #[link Layer 2 PR number]. Users can convert RDF triples to/from CSV, and convert between RDF triple and quad formats, using --format extended with --graph-name and --base-uri.

**Related Issues**
Issue #60 

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update
- [ ] Housekeeping

## Checklist:
- [x] My code follows the [ruff code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
    - [x] `poetry run pytest` - all tests passed
    - [x] `poetry run ruff check` - no linting errors
    
## What was added:

- databusclient/filehandling/mapping.py — 5 mapping functions using the Layer 2 handlers as IR (TripleHandler/QuadHandler/TSDHandler). Companion .meta.json preserves RDF datatypes and language tags for lossless CSV round trips. Blank nodes serialized with _: prefix for correct round-trip reconstruction.
- --graph-name and --base-uri flags added to download command.

## Tests:

- 5 mapping round trip tests in tests/test_format_round_trips.py (IR captured before conversion, consistent with Layer 2 pattern)
- 29 functional tests in tests/test_mapping_conversions.py covering all 5 directions plus edge cases (blank nodes, missing companion file, missing required flags)

Closes #60

### CODE REVIEW

File: tests/test_mapping_conversions.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+    ex:member <alice> .
+"""
+
+SAMPLE_NQ_CONTENT = """\
+<https://example.org/data/alice> <http://xmlns.com/foaf/0.1/name> "Alice" <https://example.org/graph/people> .
+<https://example.org/data/alice> <https://example.org/vocab/age> "29"^^<http://www.w3.org/2001/XMLSchema#integer> <https://example.org/graph/people> .
+<https://example.org/data/bob> <http://xmlns.com/foaf/0.1/name> "Bob" <https://example.org/graph/people> .
+<https://example.org/data/project1> <https://example.org/vocab/title> "Databus Example Project" <https://example.org/graph/projects> .
+<https://example.org/data/project1> <https://example.org/vocab/member> <https://example.org/data/alice> <https://example.org/graph/projects> .
+"""
```
Comment (Integer-Ctrl) [inline tests/test_mapping_conversions.py]: Please put test data in `tests/resources`

File: tests/test_mapping_conversions.py
Comment (Integer-Ctrl) [inline tests/test_mapping_conversions.py:1]: We should align these tests more closely with the round-trip mapping idea from the paper:

> Within a round trip mapping test, we take a ﬁle i and convert it to ﬁle c of
the format of another equivalence class before we convert c back to ﬁle o of the
same format as i. Therefore, i ﬁrst has to be read into the internal data structure
of the equivalence class of i (see (1) in Fig. 3). Then this data is mapped to the
internal data structure of the equivalence class of c (2) before it is written out
to c (3). Next, c is read into the internal data structure of its equivalence class
(4). That resulting data is mapped back to the internal data structure of the
equivalence class i (5). In the last step (6), this internal data structure data is
written out to o. If the information of the input ﬁle i is equal to the information
of output ﬁle o, the round trip test succeeds.

The core test should not only check that each conversion direction works, files are created, or counts are non-zero. Instead, it should test:

So I would suggest adding explicit round-trip tests such as:

Triple → Quad → Triple
Triple → CSV/TSV → Triple

and compare the original RDF graph with the reconstructed RDF graph.

For mappings like Quad → Triple → Quad you have to be creative. When two files are created from one Quad file, they could be both separately converted back to Quad and merged and then compared. 

Of course a test like Quad -> TSD -> QUAD does not work since TSD to QUAD does not exist

**Review by Integer-Ctrl — CHANGES_REQUESTED** [review]

File: tests/test_mapping_conversions.py [hunk 2]
```diff
@@ -46,43 +49,6 @@ def resource(filename: str) -> str:
 quad_handler = QuadHandler()
 tsd_handler = TSDHandler()
```
Comment (Integer-Ctrl) [inline tests/test_mapping_conversions.py]:
```
triple_handler
quad_handler
tsd_handler
```
are defined twice

File: tests/test_mapping_conversions.py [hunk 3]
```diff
@@ -493,6 +457,141 @@ def test_uses_resource_files(self):
...
+                stem = os.path.basename(out_file)[:-3]  # strip .nt
+
+                # Find the matching original graph by last URI segment
+                matching_uri = next(
+                    (uri for uri in original_graphs
+                     if uri.rstrip("/").split("/")[-1] == stem),
+                    None
+                )
+                if matching_uri is None:
+                    continue
```
Comment (Integer-Ctrl) [inline tests/test_mapping_conversions.py]: This seems like an error which should raise an error instead of ignoring it with continue
E.g. `assert matching_uri is not None, f"No original graph found for {out_file}"`

File: tests/test_mapping_conversions.py (hunk 1 above)
Comment (DhanashreePetare) [reply tests/test_mapping_conversions.py]: Fixed — all test data moved to tests/resources/.  Added empty.nq, empty.csv, and missing_resource_col.csv for edge case tests.

File: tests/test_mapping_conversions.py [hunk 4]
```diff
@@ -493,6 +457,141 @@ def test_uses_resource_files(self):
...
+                "reconstructed graph is not isomorphic to original"
+            )
+
+    def test_quad_to_triple_to_quad_round_trip(self):
+        """Quad -> Triple (split) -> Quad (re-promote each graph): datasets must match.
+
+        Since Quad -> Triple produces one file per named graph, each graph is
+        separately re-promoted back to a quad Dataset and the named graphs are
+        compared individually using isomorphic().
+        """
```
Comment (Integer-Ctrl) [inline tests/test_mapping_conversions.py:553]: This is not really a round trip test because the .nt files are not converted back to .nt and compared with original dataset. I would expect something like:
```
def test_quad_to_triple_to_quad_round_trip(self):
    """Quad -> Triple files -> Quad files -> merged Quad dataset.

    The reconstructed named graphs must be isomorphic to the original named graphs.
    """
    from rdflib import Dataset, URIRef

    source = resource("sample.nq")
    d_original = quad_handler.read(source, "nquads")

    original_graphs = {
        str(g.identifier): g
        for g in d_original.graphs()
        if len(g) > 0 and str(g.identifier) not in ("urn:x-rdflib:default", "")
    }

    with tempfile.TemporaryDirectory() as tmpdir:
        # Step 1: Quad -> Triple files
        split_dir = os.path.join(tmpdir, "split")
        nt_files = convert_quads_to_triples(
            source, split_dir, "nquads", "ntriples"
        )

        assert len(nt_files) == len(original_graphs)

        used_graph_uris = set()
        nq_files = []

        # Step 2: Each Triple file -> Quad file again
        for nt_file in nt_files:
            g_split = triple_handler.read(nt_file, "ntriples")

            matching_graph_uri = None
            for graph_uri, g_original in original_graphs.items():
                if graph_uri in used_graph_uris:
                    continue
                if g_split.isomorphic(g_original):
                    matching_graph_uri = graph_uri
                    break

            assert matching_graph_uri is not None, (
                f"No matching original graph found for split file {nt_file}"
            )

            used_graph_uris.add(matching_graph_uri)

            nq_path = os.path.join(
                tmpdir,
                os.path.basename(nt_file).replace(".nt", ".nq")
            )

            convert_triples_to_quads(
                nt_file,
                nq_path,
                "ntriples",
                "nquads",
                matching_graph_uri,
            )

            nq_files.append(nq_path)

        assert used_graph_uris == set(original_graphs.keys())

        # Step 3: Merge all reconstructed NQ files into one Dataset
        d_roundtrip = Dataset()

        for nq_file in nq_files:
            d_part = quad_handler.read(nq_file, "nquads")

            for graph in d_part.graphs():
                graph_uri = str(graph.identifier)
                if graph_uri in ("urn:x-rdflib:default", "") or len(graph) == 0:
                    continue

                target_graph = d_roundtrip.graph(URIRef(graph_uri))
                for triple in graph:
                    target_graph.add(triple)

        # Step 4: Compare original named graphs with reconstructed named graphs
        roundtrip_graphs = {
            str(g.identifier): g
            for g in d_roundtrip.graphs()
            if len(g) > 0 and str(g.identifier) not in ("urn:x-rdflib:default", "")
        }

        assert set(roundtrip_graphs.keys()) == set(original_graphs.keys())

        for graph_uri, g_original in original_graphs.items():
            assert g_original.isomorphic(roundtrip_graphs[graph_uri]), (
                f"Quad -> Triple -> Quad round trip failed for graph {graph_uri}"
            )
```

File: tests/test_mapping_conversions.py
Comment (DhanashreePetare) [reply tests/test_mapping_conversions.py:1]: Fixed — added TestRoundTrips class with explicit IR comparison.

File: tests/test_mapping_conversions.py (hunk 4 above)
Comment (DhanashreePetare) [reply tests/test_mapping_conversions.py:553]: Sorry for not updating you on this, actually I was in the middle of updating this section only, the current commit is what i just did which is the correct one where I fixed this specific part.

File: tests/test_mapping_conversions.py (hunk 1 above)
Comment (Integer-Ctrl) [reply tests/test_mapping_conversions.py]: Cool. You can always click "Resolve conversation" at the bottom of a review comment. It's like a check mark on a todo list 
Screenshot from 2026-06-29 08-50-49

File: tests/test_mapping_conversions.py [hunk 5]
```diff
@@ -546,52 +546,105 @@ def test_triple_to_tsv_to_triple_round_trip_with_companion(self):
...
+                # Derive stem from filename to match back to original graph URI
                 stem = os.path.basename(out_file)[:-3]  # strip .nt
 
-                # Find the matching original graph by last URI segment
                 matching_uri = next(
                     (uri for uri in original_graphs
                      if uri.rstrip("/").split("/")[-1] == stem),
-                    None
+                    None,
+                )
```
Comment (Integer-Ctrl) [inline tests/test_mapping_conversions.py]: This part makes the round-trip test weaker than intended, because the original graph URI is recovered from the generated filename:

```python
stem = os.path.basename(out_file)[:-3]
matching_uri = next(
    (uri for uri in original_graphs
     if uri.rstrip("/").split("/")[-1] == stem),
    None,
)
```

This works for the current sample data, for example `people.nt` matching `.../people`, but it does not generally prove that `Quad -> Triple -> Quad` is lossless. Two different graph URIs could have the same last path segment, and filename sanitization may also make the mapping ambiguous.

Instead of matching by filename, we can read the generated `.nt` file and compare its triples with the original named graphs. Then we use the matching original graph URI for the re-promotion step:

```python
used_graph_uris = set()

for out_file in files:
    g_split = triple_handler.read(out_file, "ntriples")

    matching_uri = next(
        (
            uri for uri, g_original in original_graphs.items()
            if uri not in used_graph_uris
            and g_split.isomorphic(g_original)
        ),
        None,
    )

    assert matching_uri is not None, (
        f"Could not match output file '{os.path.basename(out_file)}' "
        "to any original named graph by graph content"
    )

    used_graph_uris.add(matching_uri)

    repromoted_path = os.path.join(
        tmpdir,
        f"{os.path.basename(out_file)[:-3]}_repromoted.nq"
    )

    convert_triples_to_quads(
        out_file,
        repromoted_path,
        "ntriples",
        "nquads",
        matching_uri,
    )

    # then read repromoted_path and merge as before
```

This makes the test independent from the filename convention and checks the actual information content of each split graph.

I would also add a final check that the merged dataset contains exactly the original named graph URIs and no additional graphs:

```python
merged_graph_uris = {
    str(g.identifier)
    for g in d_merged.graphs()
    if len(g) > 0 and str(g.identifier) not in ("urn:x-rdflib:default", "")
}

assert merged_graph_uris == set(original_graphs.keys())
```

File: tests/test_mapping_conversions.py (hunk 3 above)
Comment (DhanashreePetare) [reply tests/test_mapping_conversions.py]: Updated in current commit.

File: tests/test_mapping_conversions.py (hunk 2 above)
Comment (DhanashreePetare) [reply tests/test_mapping_conversions.py]: Fixed.

**Comment by Integer-Ctrl** [comment]
Please use **Resolve conversation** on the review comments when done

Oh, what a timing :smiley: 

Then, on my part, I commented on everything I had in mind. So you can proceed merging when you feel ready

File: tests/test_mapping_conversions.py (hunk 5 above)
Comment (DhanashreePetare) [reply tests/test_mapping_conversions.py]: Updated — test_quad_to_triple_to_quad_round_trip now matches each split .nt file to its original named graph by content (isomorphic() comparison) rather than filename. used_graph_uris prevents duplicate matching. Final assertion verifies merged Dataset contains exactly the original named graph URIs. Sharing the updated method for a final review before resolving conversations and merging. A request to once have a look on this function before I proceed.
```
    def test_quad_to_triple_to_quad_round_trip(self):
        """Quad -> Triple (split) -> Quad (re-promote + merge) -> compare with original.

        Follows the paper's round trip pattern (Fig. 3, steps 1-6).
        Each split .nt file is matched to its original named graph by content
        (isomorphic comparison), not by filename, making the test independent
        of filename conventions and sanitization.

        After re-promotion and merging, the merged Dataset is verified to
        contain exactly the original named graph URIs — no more, no less.
        """
        from rdflib import Dataset, URIRef

        source = resource("sample.nq")
        d_original = quad_handler.read(source, "nquads")

        # Collect original named graphs (excluding empty default graph)
        original_graphs = {
            str(g.identifier): g
            for g in d_original.graphs()
            if len(g) > 0
            and str(g.identifier) not in ("urn:x-rdflib:default", "")
        }

        assert len(original_graphs) >= 1, (
            "sample.nq must contain at least one named graph for this test"
        )

        with tempfile.TemporaryDirectory() as tmpdir:
            # Steps 2+3: Quad -> Triple (split into one .nt file per named graph)
            output_dir = os.path.join(tmpdir, "split")
            files = convert_quads_to_triples(
                source, output_dir, "nquads", "ntriples"
            )

            assert len(files) == len(original_graphs), (
                f"Expected {len(original_graphs)} output file(s) "
                f"(one per named graph), got {len(files)}"
            )

            # Steps 4+5+6: Match each .nt to original graph by content,
            # re-promote to Quad using original graph URI, merge all
            d_merged = Dataset()
            used_graph_uris = set()

            for out_file in files:
                # Step 4: Read split .nt into IR
                g_split = triple_handler.read(out_file, "ntriples")

                # Step 5: Match to original named graph by content (not filename)
                matching_uri = next(
                    (
                        uri for uri, g_original in original_graphs.items()
                        if uri not in used_graph_uris
                        and g_split.isomorphic(g_original)
                    ),
                    None,
                )
                assert matching_uri is not None, (
                    f"Could not match output file '{os.path.basename(out_file)}' "
                    "to any original named graph by graph content"
                )
                used_graph_uris.add(matching_uri)

                # Step 5+6: Re-promote .nt back to Quad using matched graph URI
                stem = os.path.basename(out_file)[:-3]
                repromoted_path = os.path.join(tmpdir, f"{stem}_repromoted.nq")
                convert_triples_to_quads(
                    out_file,
                    repromoted_path,
                    "ntriples",
                    "nquads",
                    matching_uri,
                )

                # Read repromoted Quad into IR and merge into d_merged
                d_repromoted = quad_handler.read(repromoted_path, "nquads")
                for named_graph in d_repromoted.graphs():
                    graph_id = str(named_graph.identifier)
                    if (
                        graph_id in ("urn:x-rdflib:default", "")
                        or len(named_graph) == 0
                    ):
                        continue
                    merged_graph = d_merged.graph(URIRef(graph_id))
                    for triple in named_graph:
                        merged_graph.add(triple)

            # Verify merged Dataset contains exactly the original named graph URIs
            merged_graph_uris = {
                str(g.identifier)
                for g in d_merged.graphs()
                if len(g) > 0
                and str(g.identifier) not in ("urn:x-rdflib:default", "")
            }
            assert merged_graph_uris == set(original_graphs.keys()), (
                f"Merged Dataset graph URIs do not match original. "
                f"Expected: {set(original_graphs.keys())}, "
                f"got: {merged_graph_uris}"
            )

            # Compare each named graph in d_merged against d_original
            for uri, g_original_named in original_graphs.items():
                g_merged_named = d_merged.get_context(URIRef(uri))
                assert g_original_named.isomorphic(g_merged_named), (
                    f"Quad -> Triple -> Quad round trip failed for graph '{uri}': "
                    f"reconstructed graph is not isomorphic to original. "
                    f"Original had {len(g_original_named)} triple(s), "
                    f"reconstructed has {len(g_merged_named)} triple(s)."
                )
```

Comment (Integer-Ctrl) [reply tests/test_mapping_conversions.py]: From my side it is fine to proceed

Comment (DhanashreePetare) [reply tests/test_mapping_conversions.py]: Yepp! Thanks for review.

**Review by DhanashreePetare — COMMENTED** [review]
Resolved.
